### ISSUE TITLE
Final average accuracy not printed in logs

### ISSUE THREAD

**Issue by markmcd-aiml** [sun-hailong/LAMDA-PILOT#9]
Hi @sun-hailong, 

At the end of a run, the metrics (in this case for exps/simplecil_inr.json) are printed out like this:

```
2024-01-18 10:00:30,251 [trainer.py] => No NME accuracy.
2024-01-18 10:00:30,251 [trainer.py] => CNN: {'total': 61.28, '00-19': 60.96, '20-39': 61.39, '40-59': 62.38, '60-79': 63.4, '80-99': 56.06, '100-119': 60.45, '120-139': 63.44, '140-159': 62.41, '160-179': 61.3, '180-199': 60.1, 'old': 61.41, 'new': 60.1}
2024-01-18 10:00:30,251 [trainer.py] => CNN top1 curve: [78.96, 72.14, 70.11, 68.29, 66.0, 64.42, 64.05, 63.08, 62.28, 61.28]
2024-01-18 10:00:30,251 [trainer.py] => CNN top5 curve: [95.36, 90.01, 86.4, 84.07, 82.35, 80.8, 79.9, 78.94, 77.75, 76.68]

Average Accuracy (CNN): 67.061
2024-01-18 10:00:30,251 [trainer.py] => Average Accuracy (CNN): 67.061 

Accuracy Matrix (CNN):
[[78.96 75.33 71.41 68.94 67.2  65.02 64.59 62.84 61.39 60.96]
 [ 0.   68.67 66.46 65.66 65.35 64.56 63.77 62.5  62.34 61.39]
 [ 0.    0.   72.44 69.64 66.83 65.68 65.02 64.69 63.37 62.38]
 [ 0.    0.    0.   69.   67.78 65.85 65.5  64.97 64.1  63.4 ]
 [ 0.    0.    0.    0.   62.01 60.78 59.34 57.7  56.88 56.06]
 [ 0.    0.    0.    0.    0.   63.94 63.18 61.67 61.06 60.45]
 [ 0.    0.    0.    0.    0.    0.   66.26 65.55 64.15 63.44]
 [ 0.    0.    0.    0.    0.    0.    0.   63.92 63.37 62.41]
 [ 0.    0.    0.    0.    0.    0.    0.    0.   63.18 61.3 ]
 [ 0.    0.    0.    0.    0.    0.    0.    0.    0.   60.1 ]]
2024-01-18 10:00:30,251 [trainer.py] => Forgetting (CNN): 6.2877777777777775
```
The issue is that none of the printed/logged metrics include "Final average accuracy", which is the most often used CIL metric. 

- The printed "Average Accuracy" is the average of the final CNN top 1 curve. But this is defined quite differently to "Final average accuracy"
- The "Final Average Accuracy" is easy to compute from the information you do print: its just the average of your task split values in the CNN dict, or equivalently the average of the final column in "Accuracy Matrix (CNN)"
- In this example the Final Average Accuracy calculated from the accuracy matrix is 61.189%
     -  often its close to the last value in the top-1 curve, but not always, depending on the dataset

It would be appreciated if you could change what gets printed out so that comparisons with values reported in the literature are much more easily made.

As an aside, "Accuracy Matrix (CNN)" is only printed to the terminal, and not logged to the log file. It would be good to have it in the log file.

**Comment by sun-hailong** [comment]
Hi @markmcd-aiml ,
Thanks for your careful observation. 
In our framework, the "Final average accuracy" is the last value of the CNN top1 curve list, which is 61.28% in your example. Be carefully, the value is not 61.28. Because the dataset does not have a perfectly balanced number of instances for each category, so it is not a straightforward average for the last column in the Accuracy Matrix (CNN).

**Comment by markmcd-aiml** [comment]
@sun-hailong I think I do not agree that the last value in the CNN top 1 curve list is generally equal to the final average accuracy. 

A convenient reference for the definition of "Final Average Accuracy" is the DualPrompt paper, on page 20. Here is a screenshot:

<img width="652" alt="AverageAccuracy" src="https://github.com/sun-hailong/LAMDA-PILOT/assets/125258386/40d327c6-f121-4848-88b4-e45d0d0df0f9">

By inspecting your code, I concluded that the S_{t,\tau} terms are the entries in your Accuracy matrix, and for t=10 we have S_{10,1}...S_{10,10} equal to the final "CNN" values: '00-19': 60.96, '20-39': 61.39, '40-59': 62.38, '60-79': 63.4, '80-99': 56.06, '100-119': 60.45, '120-139': 63.44, '140-159': 62.41, '160-179': 61.3, '180-199': 60.1 

According to the definition, in the screen shot, Final Average Accuracy is the average of these 10 values.

On the other hand, your Top 1 accuracy is the accuracy of the model after 10 tasks on all data points. 

Equality will follow between this Top 1 accuracy and Final Average Accuracy, if all tasks have equal numbers of samples, but it won't be exact otherwise.  Suppose in one task you have 10 samples and accuracy is 1%, and the other tasks have 1000 samples and accuracy 99%... Your top 1 will be close to 99%, but average accuracy over tasks will be close to only 90%....

**Comment by YingjianLi** [comment]
Hi, @markmcd-aiml, did you try the 'l2p_inr.json'? I found the result calculated from Accuracy Matrix (CNN) is about 72+% (https://github.com/sun-hailong/LAMDA-PILOT/issues/7), which is much higher than the result from original paper dualprompt. Do you any idea about this question? 

Thank you and have a nice day!

**Comment by markmcd-aiml** [comment]
Hi @YingjianLi ,  This difference is not because of the accuracy definition. Both the final top 1 and the final average accuracy are 72+%.  This same result of ~72% for L2P is also reported in the CODA-Prompt paper and in Table 22 in Zhou et al (2023) (https://arxiv.org/abs/2303.07338). 

One thingI'm not sure of is whether the original L2P paper used a different backbone checkpoint?  I note also in Table 22 of Zhou et al (2023) that a different backbone gives 69.77% for L2P, so it matter a lot what the checkpoint is.

Also, have a look at the CODA-Prompt paper - they also report 69% for "L2P++" and 71.66 for "Deep L2P++" - I haven't studied what they did to get the ++ but there are comments in the paper about how they improved the original L2P.  So my second guess is that the PILOT repo implements the improved method.

**Comment by YingjianLi** [comment]
@markmcd-aiml  thank you! I will check it.

**Comment by sun-hailong** [comment]
Hi @markmcd-aiml ,
We primarily focus on class-incremental learning, and for this setting, our evaluation metric follows the iCaRL paper(https://arxiv.org/abs/1611.07725). Furthermore, I believe we should not solely focus on the average accuracy of all tasks. For example, if task 1 has 100 samples with 0% accuracy, and task 2 has 1 sample with 100% accuracy, according to the metric you mentioned, the Final Average Accuracy would be 50%, which doesn't seem quite right either.
In conclusion, for class-incremental learning, I believe that in each test phase, we should not test by task but instead test all classes together, calculating the top-1 accuracy for all classes combined, and then get the task accuracy based on this overall accuracy.

**Comment by markmcd-aiml** [comment]
@sun-hailong Thanks for the answer.
- I see that what you print as "Average Accuracy" is what iCaRL calls "average incremental accuracy". This means me think it would have been better if Final Average Accuracy had been called "Final Per-task Average Accuracy," or something like that, as "Average Accuracy" makes it easy to be confused between the two.
- I definitely think you have a valid point that the final top-1 accuracy might be more useful than the Final Average Accuracy definition.  
- But I don't  have a strong view on what to choose as the metric - the choice is not the purpose of this discussion. My reasons for raising this issue is only about (i) removing ambiguity and (ii) ensuring comparisons can be easily made between PILOT results and published papers, using  consistent metrics. Many recent papers (or at least the ones I know best) use Final Average Accuracy but not Top-1.
- To this end, I have a minimal change suggestion:
1. print the words "Average incremental accuracy" instead of "Average Accuracy" 
2. additionally print a line called "CNN Per-Task Average Accuracy Curve", which will be the average of the non-empty parts of the columns in the Accuracy Matrix.  These will often be identical to the Top-1 curve, but not always. I do exactly this in the RanPAC repository. For example, I print this (but intend to change "Ave Acc curve:" to "Per task Ave Acc curve:"):
```
2024-01-18 20:01:27,618 [trainer.py] => Group Accuracies: {'00-19': 75.04, '20-39': 75.0, '40-59': 75.25, '60-79': 75.48, '80-99': 72.9, '100-119': 78.33, '120-139': 76.98, '140-159': 80.8, '160-179': 75.52, '180-199': 73.4}
2024-01-18 20:01:27,629 [trainer.py] => Top1 curve: [88.39, 85.47, 83.24, 81.87, 79.93, 78.63, 78.12, 77.6, 77.14, 76.05]
2024-01-18 20:01:27,629 [trainer.py] => Ave Acc curve: [88.4, 85.4, 83.2, 81.8, 79.8, 78.5, 78.0, 77.4, 76.9, 75.9]
```

**Comment by sun-hailong** [comment]
@markmcd-aiml Thank you for the detailed explanation. We will consider incorporating this metric in the future. Appreciate you pointing out and discussing this issue.
